design-system/src/theme-provider.tsx

```diff
@@ -54,41 +30,79 @@ const validateTheme = (themeName?: string): ThemeName => {
   return 'default';
 };
 
-const ThemeProvider = (props: ThemeProviderProps) => {
-  const root = useRef<HTMLElement>(document.querySelector(':root'));
-  const localScopeElement = useRef<HTMLDivElement>(null);
-  const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
+// used to cover SSR builds (for instance in Gatsby)
+const isBrowser = typeof window !== 'undefined';
 
-  const changeTheme = useCallback((newTheme: string) => {
-    setTheme(validateTheme(newTheme));
-  }, []);
+const defaultParentSelector = (): HTMLElement | null =>
+  document.querySelector(':root');
 
-  useLayoutEffect(() => {
-    const vars = toVars(themes[theme]);
-    const targetElement =
-      props.scope === 'local' ? localScopeElement.current : root.current;
-    Object.entries(vars).forEach(([key, value]) => {
-      targetElement?.style.setProperty(key, value);
-    });
-  }, [theme, props.scope]);
+type TApplyTheme = {
+  newTheme?: string;
+  parentSelector: typeof defaultParentSelector;
+  themeOverrides?: Record<string, string>;
+};
 
-  const value = useMemo(() => {
-    return { theme, changeTheme };
-  }, [theme, changeTheme]);
+const applyTheme = ({
+  newTheme,
+  parentSelector = defaultParentSelector,
+  themeOverrides,
+}: TApplyTheme): void => {
+  const target = isBrowser ? parentSelector() : null;
+  const validTheme = validateTheme(newTheme);
+  target?.setAttribute('data-theme', validTheme);
+  const vars = toVars(
+    themeOverrides && isObject(themeOverrides)
+      ? merge(cloneDeep(themes[validTheme]), themeOverrides)
```

Comment:
I was wondering if we do actually need to use `cloneDeep` here. I guess the problem was that we were mutating the theme in `themes` but I guess we can also prevent from that without cloning using a different target as the merge destination:
```suggestion
      ? merge({}, themes[validTheme], themeOverrides)
```

What do you think?